src/main/java/org/broadinstitute/hellbender/tools/dataflow/transforms/bqsr/BaseRecalibratorOptimizedFn.java

```diff
@@ -0,0 +1,572 @@
+package org.broadinstitute.hellbender.tools.dataflow.transforms.bqsr;
+
+import com.google.cloud.dataflow.sdk.values.PCollectionView;
+import com.google.common.annotations.VisibleForTesting;
+import htsjdk.samtools.CigarElement;
+import htsjdk.samtools.SAMFileHeader;
+import htsjdk.samtools.SAMSequenceDictionary;
+import htsjdk.samtools.util.Locatable;
+import org.broadinstitute.hellbender.engine.ReferenceDataSource;
+import org.broadinstitute.hellbender.engine.ReferenceMemorySource;
+import org.broadinstitute.hellbender.engine.dataflow.DoFnWLog;
+import org.broadinstitute.hellbender.engine.dataflow.datasources.ReadContextData;
+import org.broadinstitute.hellbender.engine.dataflow.transforms.composite.AddContextDataToReadOptimized;
+import org.broadinstitute.hellbender.exceptions.GATKException;
+import org.broadinstitute.hellbender.tools.recalibration.QuantizationInfo;
+import org.broadinstitute.hellbender.tools.recalibration.ReadCovariates;
+import org.broadinstitute.hellbender.tools.recalibration.RecalUtils;
+import org.broadinstitute.hellbender.tools.recalibration.RecalibrationTables;
+import org.broadinstitute.hellbender.tools.recalibration.covariates.StandardCovariateList;
+import org.broadinstitute.hellbender.tools.walkers.bqsr.ReadRecalibrationInfo;
+import org.broadinstitute.hellbender.tools.walkers.bqsr.RecalibrationEngine;
+import org.broadinstitute.hellbender.transformers.ReadTransformer;
+import org.broadinstitute.hellbender.utils.BaseUtils;
+import org.broadinstitute.hellbender.utils.MathUtils;
+import org.broadinstitute.hellbender.utils.baq.BAQ;
+import org.broadinstitute.hellbender.utils.clipping.ReadClipper;
+import org.broadinstitute.hellbender.utils.read.AlignmentUtils;
+import org.broadinstitute.hellbender.utils.read.GATKRead;
+import org.broadinstitute.hellbender.utils.read.ReadUtils;
+import org.broadinstitute.hellbender.utils.read.SAMRecordToGATKReadAdapter;
+import org.broadinstitute.hellbender.utils.recalibration.EventType;
+import org.broadinstitute.hellbender.utils.reference.ReferenceBases;
+import org.broadinstitute.hellbender.utils.variant.Variant;
+
+import java.io.File;
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.io.PrintStream;
+import java.util.Arrays;
+
+/**
+ * DoFn for BaseRecalibrator.
+ * Takes in reads + contextual data (overlapping reference bases and variants), spits out RecalibrationTables.
+ */
+public final class BaseRecalibratorOptimizedFn extends DoFnWLog<AddContextDataToReadOptimized.ContextShard, RecalibrationTables> {
```

Comment:
According to my diff tool, this class is almost identical to `BaseRecalibratorFn` above -- when we rebase https://github.com/broadinstitute/hellbender/tree/dr_lb_spark_bqsr on top of this, `BaseRecalibratorFn` and `BaseRecalibratorOptimizedFn` will both become thin shells that call into a shared `BaseRecalibrationEngine`. If there's any subtle optimization in this class that I'm not seeing, please let me know so that I can preserve it during the rebase.